### PR TITLE
Create new user

### PR DESCRIPTION
**Pull request by nashdatle**

### CODE REVIEW

File: backend/API/appsettings.Development.json [hunk 1]
```diff
@@ -6,6 +6,6 @@
     }
   },
     "ConnectionStrings": {
-        "RookiesConnectionString": "Server=localhost\\SQLExpress;Database=TheRookiesDB;Trusted_Connection=True;"
+        "RookiesConnectionString": "Server=localhost\\MSSQLSERVER2019;Database=TheRookiesDB;Trusted_Connection=True;"
```
Comment (truongtrinhv) [inline backend/API/appsettings.Development.json]: Keep this changes as local file. No check in

File: backend/Application/DTOs/Users/CreateUser/CreateUserRequest.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    public Genders Gender { get; set; }
+
+    [Required]
+    public DateTime JoinedDate { get; set; }
+
+    [Required]
+    public UserRoles Role { get; set; }
+
+    [Required]
+    public Locations Location { get; set; }
```
Comment (truongtrinhv) [inline backend/Application/DTOs/Users/CreateUser/CreateUserRequest.cs]: Enum convension ✔️ DO use a singular type name for an enumeration unless its values are bit fields.

https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-classes-structs-and-interfaces#naming-enumerations
UserRoles, Locations, Genders must be : UserRole,Location,Gender

File: backend/Application/DTOs/Users/CreateUser/CreateUserResponse.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    public Guid Id { get; }
+
+    public string? StaffCode { get; }
+
+    public string? FirstName { get; }
+
+    public string? LastName { get; }
+
+    public string? Username { get; }
```
Comment (truongtrinhv) [inline backend/Application/DTOs/Users/CreateUser/CreateUserResponse.cs]: Property Naming convention: UserName

File: backend/Application/DTOs/Users/CreateUser/CreateUserResponse.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    public string? HashedPassword { get; }
```
Comment (cowienduckie) [inline backend/Application/DTOs/Users/CreateUser/CreateUserResponse.cs]: User's props are string. No need to use string?

Comment (cowienduckie) [inline backend/Application/DTOs/Users/CreateUser/CreateUserResponse.cs]: Dont return HashedPassword and unused props

File: backend/Application/Services/UserService.cs [hunk 5]
```diff
@@ -78,6 +83,65 @@ public async Task<Response> ChangePasswordAsync(ChangePasswordRequest requestMod
...
 
+    public async Task<Response<CreateUserResponse>> CreateUserAsync(CreateUserRequest requestModel)
+    {
+        var userRepository = UnitOfWork.AsyncRepository<User>();
+
+        var user = new User();
+
+        var responseModel = new CreateUserResponse(user);
+
+        if (GetAge(requestModel.DateOfBirth) < 18)
```
Comment (truongtrinhv) [inline backend/Application/Services/UserService.cs]: This code should be seperate as local variable

File: backend/Application/Services/UserService.cs [hunk 6]
```diff
@@ -78,6 +83,65 @@ public async Task<Response> ChangePasswordAsync(ChangePasswordRequest requestMod
...
+        {
+            return new Response<CreateUserResponse>(false, ErrorMessages.InvalidAge, responseModel);
+        }
+
+        if (DateTime.Compare(requestModel.DateOfBirth, requestModel.JoinedDate) != 1
```
Comment (truongtrinhv) [inline backend/Application/Services/UserService.cs]: Same with above comment

File: backend/Application/Services/UserService.cs [hunk 7]
```diff
@@ -78,6 +83,65 @@ public async Task<Response> ChangePasswordAsync(ChangePasswordRequest requestMod
...
+            || requestModel.JoinedDate.DayOfWeek == DayOfWeek.Saturday || requestModel.JoinedDate.DayOfWeek == DayOfWeek.Sunday)
```
Comment (truongtrinhv) [inline backend/Application/Services/UserService.cs]: Break this check to short line
DateTime.Compare(requestModel.DateOfBirth, requestModel.JoinedDate) != 1
            || requestModel.JoinedDate.DayOfWeek == DayOfWeek.Saturday 
            || requestModel.JoinedDate.DayOfWeek == DayOfWeek.Sunday

File: backend/Application/Services/UserService.cs [hunk 8]
```diff
@@ -78,6 +83,65 @@ public async Task<Response> ChangePasswordAsync(ChangePasswordRequest requestMod
...
+        {
+            return new Response<CreateUserResponse>(false, ErrorMessages.InvalidJoinedDate, responseModel);
+        }
+
+        var latestStaffCode = userRepository.ListAsync(user => user.IsDeleted == false).Result.OrderByDescending(user => user.StaffCode).First().StaffCode;
```
Comment (truongtrinhv) [inline backend/Application/Services/UserService.cs]: Wrap this code to short line code

File: backend/Application/Services/UserService.cs [hunk 9]
```diff
@@ -78,6 +83,65 @@ public async Task<Response> ChangePasswordAsync(ChangePasswordRequest requestMod
...
+
+        if (latestStaffCode == null)
+        {
+            return new Response<CreateUserResponse>(false, ErrorMessages.InternalServerError, responseModel);
+        }
+
+        var latestUserName = userRepository.ListAsync(user => user.IsDeleted == false).Result.OrderByDescending(user => user.StaffCode).First().Username;
```
Comment (truongtrinhv) [inline backend/Application/Services/UserService.cs]: Same with above comment. Wrap this code to short line

File: backend/Application/Services/UserService.cs [hunk 10]
```diff
@@ -119,7 +183,7 @@ public async Task<Response<GetListUsersResponse>> GetListAsync(GetListUsersReque
                                 .Select(u => new GetUserResponse(u))
                                 .AsQueryable();
 
-        var validSortFields = new []
+        var validSortFields = new[]
```
Comment (truongtrinhv) [inline backend/Application/Services/UserService.cs]: Format this code

File: backend/Application/Services/UserService.cs [hunk 11]
```diff
@@ -156,4 +220,58 @@ public async Task<Response<GetListUsersResponse>> GetListAsync(GetListUsersReque
...
     }
+
+    public static int GetAge(DateTime birthDate)
+    {
+        var today = DateTime.Now;
+
+        var age = today.Year - birthDate.Year;
+
+        if (today.Month < birthDate.Month || (today.Month == birthDate.Month && today.Day < birthDate.Day)) { age--; }
+
```
Comment (truongtrinhv) [inline backend/Application/Services/UserService.cs:239]: Wrap the code. Each line would be for one condition

File: backend/Application/Services/UserService.cs [hunk 12]
```diff
@@ -156,4 +220,58 @@ public async Task<Response<GetListUsersResponse>> GetListAsync(GetListUsersReque
...
+        return age;
+    }
+
+    public static string GetNewStaffCode(string previousStaffCode)
+    {
+        var prefix = "SD";
```
Comment (truongtrinhv) [inline backend/Application/Services/UserService.cs]: Should be move to the constant file

File: backend/Domain/Shared/Constants/ErrorMessages.cs [hunk 13]
```diff
@@ -7,6 +7,10 @@ public static class ErrorMessages
     public const string Unauthorized = "Unauthorized";
     public const string NotFound = "Not Found";
     public const string BadRequest = "Bad request!";
+
```
Comment (truongtrinhv) [inline backend/Domain/Shared/Constants/ErrorMessages.cs]: Remove empty line

File: backend/Domain/Shared/Constants/ErrorMessages.cs [hunk 14]
```diff
@@ -7,6 +7,10 @@ public static class ErrorMessages
...
     public const string MatchingOldAndNewPassword = "The new password cannot match the old password!";
     public const string WrongOldPassword = "Old password is wrong!";
+
```
Comment (cowienduckie) [inline backend/Domain/Shared/Constants/ErrorMessages.cs]: Remove empty line

**Review by cowienduckie — APPROVED** [review]

**Pull request merged**